### PR TITLE
fix(tts): don't pad an already-frame-aligned audio target with a spurious frame

### LINKED ISSUE #16249: MagpieTTS raw-audio padding adds a spurious silent frame to exact-multiple-length audio
**Describe the bug**

`MagpieTTSDataset.__getitem__` and `MagpieTTSLhotseDataset`'s equivalent raw-audio branch pad loaded audio up to a multiple of `codec_model_samples_per_frame` with:

```python
audio = torch.nn.functional.pad(
    audio,
    (0, self.codec_model_samples_per_frame - (audio.shape[0] % self.codec_model_samples_per_frame)),
    value=0,
)
```

When `audio.shape[0]` is already an exact multiple of `codec_model_samples_per_frame`, the modulo is `0` and the pad amount becomes `codec_model_samples_per_frame` itself instead of `0` — a full spurious silent frame is appended to the training target audio.

This branch is the fallback whenever a manifest entry has no pre-computed `target_audio_codes_path` (or `load_cached_codes_if_available` is `False`), which is a real, documented path for raw-audio MagpieTTS training, not a corner case.

**Steps/Code to reproduce bug**

```python
import torch

codec_model_samples_per_frame = 256
audio = torch.zeros(codec_model_samples_per_frame * 5)  # already an exact multiple

padded = torch.nn.functional.pad(
    audio,
    (0, codec_model_samples_per_frame - (audio.shape[0] % codec_model_samples_per_frame)),
    value=0,
)
print(padded.shape[0])  # 1536, expected 1280 (no padding needed)
```

**Expected behavior**

Audio whose length already divides `codec_model_samples_per_frame` evenly should not be padded at all (`audio_len` unchanged), matching the sibling guard in `nemo/collections/tts/data/dataset.py`'s `_pad_wav_to_multiple`, which explicitly checks `if wav.shape[0] % self.pad_multiple != 0` before padding.

**Environment overview**

- Environment location: Bare-metal
- Method of NeMo install: from source (`main` branch)

**Environment details**

- OS version: Ubuntu 24.04
- PyTorch version: 2.13.0+cpu
- Python version: 3.12.3

**Additional context**

Both occurrences share the identical formula:
- `nemo/collections/tts/data/text_to_speech_dataset.py` (`MagpieTTSDataset.__getitem__`)
- `nemo/collections/tts/data/text_to_speech_dataset_lhotse.py` (`MagpieTTSLhotseDataset`)

Introduced in `19367c9c25` (#15031, "Update MagpieTTS model with latest changes"); not exercised by any existing test.

### PR DESCRIPTION
**Pull request by udsy19**
# What does this PR do ?

Fix a spurious extra padding frame appended to already-frame-aligned raw audio targets in `MagpieTTSDataset` and `MagpieTTSLhotseDataset`.

**Collection**: tts

# Changelog

- `nemo/collections/tts/data/text_to_speech_dataset.py`: `MagpieTTSDataset.__getitem__`'s raw-audio padding now uses `-audio.shape[0] % self.codec_model_samples_per_frame`, which is `0` when the audio already divides the frame size evenly, instead of `self.codec_model_samples_per_frame - (audio.shape[0] % self.codec_model_samples_per_frame)`, which produced a full spurious frame in that case.
- `nemo/collections/tts/data/text_to_speech_dataset_lhotse.py`: identical fix to `MagpieTTSLhotseDataset`'s equivalent branch.
- `tests/collections/tts/data/test_text_to_speech_dataset_padding.py`: new regression tests covering both the exact-multiple case (must not pad) and the non-multiple case (must still round up).

# Usage

```python
import torch

frame = 256
audio = torch.zeros(frame * 5)  # already an exact multiple of the frame size
padded = torch.nn.functional.pad(audio, (0, -audio.shape[0] % frame), value=0)
assert padded.shape[0] == frame * 5  # no spurious frame added
```

# GitHub Actions CI

n/a

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA-NeMo/Speech/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?

**PR Type**:

- [x] Bugfix

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.

# Additional Information

Fixes #16249

## Verification

Negative control: reverting only `text_to_speech_dataset.py` and `text_to_speech_dataset_lhotse.py` to `main` and re-running the new test file (`pytest tests/collections/tts/data/test_text_to_speech_dataset_padding.py -v`) fails without the fix and passes with it:

```
1 failed, 1 passed
```

```
2 passed
```

The `test_non_multiple_length_audio_is_rounded_up` case passes both before and after, confirming the fix only removes the spurious padding on the exact-multiple case rather than changing the general rounding behavior. Ran black (line length 119, `skip_string_normalization`), isort (`profile=black`), and flake8 on all three changed/added files with no findings.

Signed-off-by: Udaya Tejas <udayatejas2004@gmail.com>

### CODE REVIEW

**Review by blisc — APPROVED** [review]

**Review by blisc — DISMISSED** [review]

File: tests/collections/tts/data/test_text_to_speech_dataset_padding.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+# limitations under the License.
+
+from pathlib import Path
+
+import numpy as np
+import pytest
+import soundfile as sf
+
+import nemo.collections.tts.data.text_to_speech_dataset as text_to_speech_dataset
+from nemo.collections.tts.data.text_to_speech_dataset import DatasetSample, MagpieTTSDataset
```
Comment (XuesongYang) [inline tests/collections/tts/data/test_text_to_speech_dataset_padding.py]: nit: the above check makes sense. how about removing this line, and calling `text_to_speech_dataset.DatasetSample` and `text_to_speech_dataset.MagpieTTSDataset`, instead?

File: nemo/collections/tts/data/text_to_speech_dataset_lhotse.py
Comment (XuesongYang) [inline nemo/collections/tts/data/text_to_speech_dataset_lhotse.py:1]: `text_to_speech_dataset_lhotse.py:346` gets the same one-line fix but nothing exercises it. Please add exact-multiple and non-multiple cases for `MagpieTTSLhotseDataset` in `tests/collections/tts/data/test_magpietts_dataset_lhotse.py`. The existing helpers make this small:

- `_dataset_kwargs()` with `load_cached_codes_if_available=False`, and drop `target_codes` from the cut's `custom` so the raw-audio branch is taken.
- `dummy_recording(..., with_data=True, sampling_rate=24000)`: `duration=0.5` → 12000 samples = 25 × 480 (exact multiple; old code would give 12480), `duration=0.51` → 12240 samples (non-multiple; both give 12480). `Recording.resample` is a no-op at the same rate, so the sample counts are exact.
- Inject `_FakeTextTokenizer` the way the multiturn tests do, so the test stays offline and doesn't need the Nemotron tokenizer.
- Assert `batch["audio_lens"]`.

File: tests/collections/tts/data/test_text_to_speech_dataset_padding.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+def _write_wav(path: Path, num_samples: int) -> None:
+    sf.write(str(path), np.zeros(num_samples, dtype=np.float32), SAMPLE_RATE, format="WAV")
+
+
+def _make_dataset(audio_dir: Path, monkeypatch) -> MagpieTTSDataset:
+    # Bypass __init__ (and the heavyweight phoneme tokenizer it wires up) and set only the
+    # attributes the raw-audio branch of __getitem__ actually reads.
+    monkeypatch.setattr(text_to_speech_dataset, "tokenize_text_with_phoneme_spans", lambda **kwargs: [1, 2, 3])
+
+    dataset = object.__new__(MagpieTTSDataset)
```
Comment (XuesongYang) [inline tests/collections/tts/data/test_text_to_speech_dataset_padding.py]: The `object.__new__` + hand-set attributes approach is brittle: if `__getitem__` starts reading a new attribute, this test fails with `AttributeError` rather than on padding. Not blocking, but a helper that builds the dataset through `__init__` with a stub tokenizer would age better.

**Review by XuesongYang — DISMISSED** [review]

**Review by blisc — DISMISSED** [review]

**Comment by blisc** [comment]
/ok to test https://github.com/NVIDIA-NeMo/Speech/pull/16250/commits/d8504cfe6b3c647fd04b1426125ed973541e4cc8

File: tests/collections/tts/data/test_magpietts_dataset_lhotse.py [hunk 3]
```diff
@@ -198,6 +198,36 @@ def _multiturn_cutset():
     return CutSet.from_cuts([cut])
 
 
+def _raw_audio_cutset(target_audio_duration, text="hello"):
+    # No "target_codes" custom field, so __getitem__ takes the raw-audio branch regardless of
+    # load_cached_codes_if_available.
+    cut = dummy_cut(
+        2,
+        duration=0.5,
+        recording=dummy_recording(2, duration=0.5, with_data=True, sampling_rate=SAMPLE_RATE),
```
Comment (XuesongYang) [inline tests/collections/tts/data/test_magpietts_dataset_lhotse.py]: IIUC, we should not hardcode `duration` here because the parameterized fixture has defined it. 
```suggestion
        duration=target_audio_duration,
        recording=dummy_recording(2, duration=target_audio_duration, with_data=True, sampling_rate=SAMPLE_RATE),
```

File: tests/collections/tts/data/test_text_to_speech_dataset_padding.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+SAMPLE_RATE = 22050
+CODEC_MODEL_SAMPLES_PER_FRAME = 256
+
+
+def _write_wav(path: Path, num_samples: int) -> None:
+    sf.write(str(path), np.zeros(num_samples, dtype=np.float32), SAMPLE_RATE, format="WAV")
+
+
+def _make_dataset(audio_dir: Path, monkeypatch) -> text_to_speech_dataset.MagpieTTSDataset:
+    # Bypass __init__ (and the heavyweight phoneme tokenizer it wires up) and set only the
```
Comment (XuesongYang) [inline tests/collections/tts/data/test_text_to_speech_dataset_padding.py:35]: identified a misleading comment:

`MagpieTTSDataset.__init__` builds no tokenizer at all. It sets `text_tokenizer = None` with the comment "Assigned in `worker_init_fn` in model file", and does the same for the phoneme tokenizer.

File: tests/collections/tts/data/test_text_to_speech_dataset_padding.py (hunk 2 above)
Comment (XuesongYang) [reply tests/collections/tts/data/test_text_to_speech_dataset_padding.py]: let's avoid using `object.__new__` + `monkeypatch`, but instead, consider instantiating the class `text_to_speech_dataset.MagpieTTSDataset` motivated by the above description. This would save the effort for the person adding the feature who has to figure out that a padding test failed for reasons unrelated to padding.

Comment (XuesongYang) [reply tests/collections/tts/data/test_text_to_speech_dataset_padding.py]: @udsy19 for example, rewrite `def _make_dataset` func like below,
```python
def _make_dataset(audio_dir: Path, num_input_samples: int) -> text_to_speech_dataset.MagpieTTSDataset:
    # load_audio reads int(duration * sample_rate) samples, so make sure the manifest duration
    # round-trips to exactly the number of samples under test.
    duration = num_input_samples / SAMPLE_RATE
    assert int(duration * SAMPLE_RATE) == num_input_samples
    manifest_path = audio_dir / "manifest.jsonl"
    manifest_path.write_text(
        json.dumps({"audio_filepath": "sample.wav", "duration": duration, "text": "hello world"}) + "\n"
    )
    dataset = text_to_speech_dataset.MagpieTTSDataset(
        dataset_meta={"unit": {"manifest_path": str(manifest_path), "audio_dir": str(audio_dir)}},
        sample_rate=SAMPLE_RATE,
        codec_model_samples_per_frame=CODEC_MODEL_SAMPLES_PER_FRAME,
        eos_id=1,
        num_audio_codebooks=8,
        volume_norm=False,
        load_cached_codes_if_available=False,
        load_16khz_audio=False,
    )
    # Normally assigned by the model's worker_init_fn; a stub keeps the test offline.
    dataset.text_tokenizer = _FakeTextTokenizer()
    return dataset
```

**Comment by blisc** [comment]
/ok to test https://github.com/NVIDIA-NeMo/Speech/pull/16250/commits/f3d69bcdd771a26d94ffba97fed6c970cb7eb920

**Review by blisc — APPROVED** [review]

**Comment by blisc** [comment]
Thanks for the bug fix!
